dfly_bench crashes on parsing redis response

cmd: [--h=<> --p=6385 --password=<> --n=1000 --c=20 --d=16 --ratio=1:3 --proactor_threads=16 --test_time=180 --pipeline=10 --key_maximum=312500100 --logtostderr --colorlogtostderr=true --colorlogtostdout=true]

```
F20250513 13:01:36.395577 71333 redis_parser.cc:95] Check failed: str.empty()
*** Check failure stack trace: ***
    @           0x4f176b  google::LogMessage::Fail()
    @           0x4f16d2  google::LogMessage::SendToLog()
    @           0x4f0edb  google::LogMessage::Flush()
    @           0x4f47b2  google::LogMessageFatal::~LogMessageFatal()
    @           0x459eb3  facade::RedisParser::Parse()
    @           0x407891  Driver::ParseRESP()
    @           0x407607  Driver::ReceiveFb()
    @           0x405c99  _ZZN6Driver7ConnectEjRKN5boost4asio2ip14basic_endpointINS2_3tcpEEEENKUlvE_clEv
    @           0x41a868  _ZSt13__invoke_implIvZN6Driver7ConnectEjRKN5boost4asio2ip14basic_endpointINS3_3tcpEEEEUlvE_JEET_St14__invoke_otherOT0_DpOT1_
    @           0x417fc0  _ZSt8__invokeIZN6Driver7ConnectEjRKN5boost4asio2ip14basic_endpointINS3_3tcpEEEEUlvE_JEENSt15__invoke_resultIT_JDpT0_EE4typeEOSB_DpOSC_
    @           0x415425  _ZSt12__apply_implIZN6Driver7ConnectEjRKN5boost4asio2ip14basic_endpointINS3_3tcpEEEEUlvE_St5tupleIJEEJEEDcOT_OT0_St16integer_sequenceImJXspT1_EEE
    @           0x41545f  _ZSt5applyIZN6Driver7ConnectEjRKN5boost4asio2ip14basic_endpointINS3_3tcpEEEEUlvE_St5tupleIJEEEDcOT_OT0_
    @           0x4154cc  _ZN4util3fb26detail15WorkerFiberImplIZN6Driver7ConnectEjRKN5boost4asio2ip14basic_endpointINS6_3tcpEEEEUlvE_JEE4run_EONS4_7context5fiberE
    @           0x412e9c  _ZZN4util3fb26detail15WorkerFiberImplIZN6Driver7ConnectEjRKN5boost4asio2ip14basic_endpointINS6_3tcpEEEEUlvE_JEEC4INS4_7context21basic_fixedsize_stackINSF_12stack_traitsEEEEESt17basic_string_viewIcSt11char_traitsIcEENS0_13FiberPriorityERKNSF_12preallocatedEOT_OSC_ENKUlONSF_5fiberEE_clESV_
    @           0x425880  _ZSt13__invoke_implIN5boost7context5fiberERZN4util3fb26detail15WorkerFiberImplIZN6Driver7ConnectEjRKNS0_4asio2ip14basic_endpointINS9_3tcpEEEEUlvE_JEEC4INS1_21basic_fixedsize_stackINS1_12stack_traitsEEEEESt17basic_string_viewIcSt11char_traitsIcEENS4_13FiberPriorityERKNS1_12preallocatedEOT_OSF_EUlOS2_E_JS2_EEST_St14__invoke_otherOT0_DpOT1_
    @           0x4227e8  _ZSt8__invokeIRZN4util3fb26detail15WorkerFiberImplIZN6Driver7ConnectEjRKN5boost4asio2ip14basic_endpointINS7_3tcpEEEEUlvE_JEEC4INS5_7context21basic_fixedsize_stackINSG_12stack_traitsEEEEESt17basic_string_viewIcSt11char_traitsIcEENS1_13FiberPriorityERKNSG_12preallocatedEOT_OSD_EUlONSG_5fiberEE_JSV_EENSt15__invoke_resultISS_JDpT0_EE4typeEST_DpOS10_
    @           0x42056c  _ZSt6invokeIRZN4util3fb26detail15WorkerFiberImplIZN6Driver7ConnectEjRKN5boost4asio2ip14basic_endpointINS7_3tcpEEEEUlvE_JEEC4INS5_7context21basic_fixedsize_stackINSG_12stack_traitsEEEEESt17basic_string_viewIcSt11char_traitsIcEENS1_13FiberPriorityERKNSG_12preallocatedEOT_OSD_EUlONSG_5fiberEE_JSV_EENSt13invoke_resultISS_JDpT0_EE4typeEST_DpOS10_
    @           0x41e3f3  _ZN5boost7context6detail12fiber_recordINS0_5fiberENS0_21basic_fixedsize_stackINS0_12stack_traitsEEEZN4util3fb26detail15WorkerFiberImplIZN6Driver7ConnectEjRKNS_4asio2ip14basic_endpointINSD_3tcpEEEEUlvE_JEEC4IS6_EESt17basic_string_viewIcSt11char_traitsIcEENS8_13FiberPriorityERKNS0_12preallocatedEOT_OSJ_EUlOS3_E_E3runEPv
    @           0x41ab6e  _ZN5boost7context6detail11fiber_entryINS1_12fiber_recordINS0_5fiberENS0_21basic_fixedsize_stackINS0_12stack_traitsEEEZN4util3fb26detail15WorkerFiberImplIZN6Driver7ConnectEjRKNS_4asio2ip14basic_endpointINSE_3tcpEEEEUlvE_JEEC4IS7_EESt17basic_string_viewIcSt11char_traitsIcEENS9_13FiberPriorityERKNS0_12preallocatedEOT_OSK_EUlOS4_E_EEEEvNS1_10transfer_tE
    @     0x7f9c99bdb207  make_fcontext
*** SIGABRT received at time=1747121496 on cpu 11 ***
PC: @     0x7f9c9967ff54  (unknown)  __pthread_kill_implementation
    @           0x574bc5         64  absl::lts_20240722::WriteFailureInfo()
    @           0x574dd2         96  absl::lts_20240722::AbslFailureSignalHandler()
    @     0x7f9c99627050       3392  (unknown)
    @     0x7f9c99626f9e         32  gsignal
    @     0x7f9c9960e942        192  abort
    @           0x4fb2a4        176  google::DumpStackTraceAndExit()
    @           0x4f176b         16  google::LogMessage::Fail()
    @           0x4f16d2        160  google::LogMessage::SendToLog()
    @           0x4f0edb         64  google::LogMessage::Flush()
    @           0x4f47b2         32  google::LogMessageFatal::~LogMessageFatal()
    @           0x459eb3       1248  facade::RedisParser::Parse()
    @           0x407891       1120  Driver::ParseRESP()
    @           0x407607        672  Driver::ReceiveFb()
    @           0x405c99         32  Driver::Connect()::{lambda()#1}::operator()()
    @           0x41a868         32  std::__invoke_impl<>()
    @           0x417fc0         32  std::__invoke<>()
    @           0x415425         32  std::__apply_impl<>()
    @           0x41545f         48  std::apply<>()
    @           0x4154cc         80  util::fb2::detail::WorkerFiberImpl<>::run_()
    @           0x412e9c         64  util::fb2::detail::WorkerFiberImpl<>::WorkerFiberImpl<>()::{lambda()#1}::operator()()
    @           0x425880         64  std::__invoke_impl<>()
    @           0x4227e8         64  std::__invoke<>()
    @           0x42056c         64  std::invoke<>()
    @           0x41e3f3         80  boost::context::detail::fiber_record<>::run()
    @           0x41ab6e         48  boost::context::detail::fiber_entry<>()
    @     0x7f9c99bdb207  (unknown)  make_fcontext

        Error Trace:    /home/abhijat/work/controlplane/dftest/tests/cluster/replication_under_load.go:140
                                                /usr/lib/golang/src/runtime/asm_amd64.s:1700
        Error:          Received unexpected error:
                        signal: aborted (core dumped)

panic: failed

goroutine 178 [running]:
github.com/dragonflydb/controlplane/dftest.(*T).FailNow(0x23af780?)
        /home/abhijat/work/controlplane/dftest/testing.go:93 +0x32
github.com/stretchr/testify/require.NoError({0x23cebb0, 0xc0001ef7c0}, {0x23b0280, 0xc000c9b2e0}, {0x0, 0x0, 0x0})
        /home/abhijat/go/pkg/mod/github.com/stretchr/testify@v1.10.0/require/require.go:1357 +0xca
github.com/dragonflydb/controlplane/dftest/tests/cluster.increaseReplicas.func1()
        /home/abhijat/work/controlplane/dftest/tests/cluster/replication_under_load.go:140 +0x4b
created by github.com/dragonflydb/controlplane/dftest/tests/cluster.increaseReplicas in goroutine 80
        /home/abhijat/work/controlplane/dftest/tests/cluster/replication_under_load.go:138 +0xff
```

--- a/src/facade/redis_parser.cc
+++ b/src/facade/redis_parser.cc
@@ -422,13 +422,18 @@ auto RedisParser::ParseArg(Buffer str) -> ResultConsumed {
   const char* s = reinterpret_cast<const char*>(str.data());
   const char* eol = reinterpret_cast<const char*>(memchr(s, '\n', str.size()));
 
-  // TODO: in client mode we still may not consume everything (see INPUT_PENDING below).
-  // It's not a problem, because we need consume all the input only in server mode.
-
   if (arg_c_ == '+' || arg_c_ == '-') {  // Simple string or error.
     DCHECK(!server_mode_);
     if (!eol) {
-      Result r = str.size() < 256 ? INPUT_PENDING : BAD_STRING;
+      // if eol is not found we should still read input as bulk string
+      cached_expr_->emplace_back(RespExpr::STRING);
+      cached_expr_->back().u = Buffer{};
+      bulk_len_ = str.length();
+      // eol is not found but if '\r' is present decrease bulk_len
+      if (s[bulk_len_ - 1] == '\r')
+        bulk_len_--;
+      state_ = BULK_STR_S;
+      Result r = str.size() < 256 ? OK : BAD_STRING;
       return {r, 0};
     }
 
@@ -479,6 +484,16 @@ auto RedisParser::ConsumeBulk(Buffer str) -> ResultConsumed {
   uint32_t consumed = 0;
   auto& bulk_str = get<Buffer>(cached_expr_->back().u);
 
+  bool extend = false;
+  // Handle split simple message or error in client mode
+  if (!server_mode_ && (arg_c_ == '+' || arg_c_ == '-') && !bulk_len_) {
+    // Search first '\r' in next partial message which ends bulk string
+    const char* s = reinterpret_cast<const char*>(str.data());
+    const char* pos = reinterpret_cast<const char*>(memchr(s, '\r', str.size()));
+    bulk_len_ = pos ? pos - s : str.size();
+    extend = true;
+  }
+
   if (str.size() >= bulk_len_) {
     consumed = bulk_len_;
     if (bulk_len_) {
@@ -487,6 +502,8 @@ auto RedisParser::ConsumeBulk(Buffer str) -> ResultConsumed {
       if (is_broken_token_) {
         memcpy(const_cast<uint8_t*>(bulk_str.end()), str.data(), bulk_len_);
         bulk_str = Buffer{bulk_str.data(), bulk_str.size() + bulk_len_};
+      } else if (extend) {
+        ExtendBulkString(Buffer(str.begin(), bulk_len_));
       } else {
         bulk_str = str.subspan(0, bulk_len_);
       }
@@ -565,6 +582,20 @@ void RedisParser::ExtendLastString(Buffer str) {
   memcpy(nb.data(), last_str.data(), last_str.size());
   memcpy(nb.data() + last_str.size(), str.data(), str.size());
   last_str = RespExpr::Buffer{nb.data(), last_str.size() + str.size()};
+  buf_stash_.back() = std::move(nb);
+}
+
+void RedisParser::ExtendBulkString(Buffer str) {
+  DCHECK(!cached_expr_->empty() && cached_expr_->back().type == RespExpr::STRING);
+
+  Buffer& bulk_str = get<Buffer>(cached_expr_->back().u);
+
+  DCHECK(bulk_str.data() == buf_stash_.back().data());
+
+  vector<uint8_t> nb(bulk_str.size() + str.size());
+  memcpy(nb.data(), bulk_str.data(), bulk_str.size());
+  memcpy(nb.data() + bulk_str.size(), str.data(), str.size());
+  bulk_str = RespExpr::Buffer{nb.data(), bulk_str.size() + str.size()};
   buf_stash_.back() = std::move(nb);
 }
 

--- a/src/facade/redis_parser.h
+++ b/src/facade/redis_parser.h
@@ -85,6 +85,7 @@ class RedisParser {
 
   void HandleFinishArg();
   void ExtendLastString(Buffer str);
+  void ExtendBulkString(Buffer str);
 
   enum State : uint8_t {
     INLINE_S,

--- a/src/facade/redis_parser_test.cc
+++ b/src/facade/redis_parser_test.cc
@@ -159,6 +159,44 @@ TEST_F(RedisParserTest, ClientMode) {
 
   ASSERT_EQ(RedisParser::OK, Parse("*3\r\n+OK\r\n$1\r\n1\r\n*2\r\n$1\r\n1\r\n$-1\r\n"));
   ASSERT_THAT(args_, ElementsAre("OK", "1", ArrLen(2)));
+
+  ASSERT_EQ(RedisParser::INPUT_PENDING, Parse("+O"));
+  EXPECT_EQ(2, consumed_);
+  ASSERT_EQ(RedisParser::INPUT_PENDING, Parse("K\r"));
+  EXPECT_EQ(2, consumed_);
+  ASSERT_EQ(RedisParser::OK, Parse("\n"));
+  ASSERT_THAT(args_, ElementsAre("OK"));
+  EXPECT_EQ(1, consumed_);
+
+  ASSERT_EQ(RedisParser::INPUT_PENDING, Parse("+OK\r"));
+  EXPECT_EQ(4, consumed_);
+  ASSERT_EQ(RedisParser::OK, Parse("\n"));
+  ASSERT_THAT(args_, ElementsAre("OK"));
+  EXPECT_EQ(1, consumed_);
+
+  ASSERT_EQ(RedisParser::INPUT_PENDING, Parse("+"));
+  EXPECT_EQ(1, consumed_);
+  ASSERT_EQ(RedisParser::INPUT_PENDING, Parse("O"));
+  EXPECT_EQ(1, consumed_);
+  ASSERT_EQ(RedisParser::INPUT_PENDING, Parse("K"));
+  EXPECT_EQ(1, consumed_);
+  ASSERT_EQ(RedisParser::INPUT_PENDING, Parse("\r"));
+  EXPECT_EQ(1, consumed_);
+  ASSERT_EQ(RedisParser::OK, Parse("\n"));
+  EXPECT_EQ(1, consumed_);
+  ASSERT_THAT(args_, ElementsAre("OK"));
+
+  ASSERT_EQ(RedisParser::INPUT_PENDING, Parse("-"));
+  EXPECT_EQ(1, consumed_);
+  ASSERT_EQ(RedisParser::OK, Parse("ERR\r\n"));
+  EXPECT_EQ(5, consumed_);
+  ASSERT_THAT(args_, ElementsAre(ErrArg("ERR")));
+
+  ASSERT_EQ(RedisParser::INPUT_PENDING, Parse("-ERR foo"));
+  EXPECT_EQ(8, consumed_);
+  ASSERT_EQ(RedisParser::OK, Parse("\r\n"));
+  EXPECT_EQ(2, consumed_);
+  ASSERT_THAT(args_, ElementsAre("ERR foo"));
 }
 
 TEST_F(RedisParserTest, Hierarchy) {

